wfe2/verify.go

```diff
@@ -18,233 +20,511 @@ import (
+	// enforceJWSAuthType()
+	key := header.JSONWebKey
+
+	// If the key isn't considered valid by go-jose return a problem immediately
+	if !key.Valid() {
+		wfe.stats.Inc("Errors.InvalidJWK", 1)
+		logEvent.AddError("JWK in request was invalid")
+		return nil, probs.Malformed("Invalid JWK in JWS header")
+	}
+
+	return key, nil
+}
+
+// lookupJWK finds a JWK associated with the Key ID present in a provided JWS,
+// returning the JWK and a pointer to the associated account, or a problem. It
+// expects that the JWS is using the embedded Key ID style of authentication
+// and does not contain an embedded JWK. Callers should have acquired the
+// provided JWS from parseJWS to ensure it has the correct number of signatures
+// present. The provided logEvent is mutated to add any observed errors.
+func (wfe *WebFrontEndImpl) lookupJWK(
+	jws *jose.JSONWebSignature,
+	ctx context.Context,
+	request *http.Request,
+	logEvent *requestEvent) (*jose.JSONWebKey, *core.Registration, *probs.ProblemDetails) {
+	// We expect the request to be using an embedded Key ID auth type and to not
+	// contain the mutually exclusive embedded JWK.
+	prob := wfe.enforceJWSAuthType(jws, embeddedKeyID)
+	if prob != nil {
+		logEvent.AddError("JWS auth type was not expected embeddedKeyID auth")
+		return nil, nil, prob
+	}
+
+	// lookupJWK is called after parseJWS() which defends against the
+	// incorrect number of signatures.
+	header := jws.Signatures[0].Header
+	accountURL := header.KeyID
+	prefix := wfe.relativeEndpoint(request, regPath)
+	accountIDStr := strings.TrimPrefix(accountURL, prefix)
+	// Convert the account ID string to an int64 for use with the SA's
+	// GetRegistration RPC
+	accountID, err := strconv.ParseInt(accountIDStr, 10, 64)
+	if err != nil {
+		wfe.stats.Inc("Errors.InvalidKeyID", 1)
+		logEvent.AddError("JWS key ID was invalid int64")
+		return nil, nil, probs.Malformed(fmt.Sprintf("Malformed account ID in KeyID header"))
+	}
+
+	// Try to find the account for this account ID
+	account, err := wfe.SA.GetRegistration(ctx, accountID)
+	// If there was an error and it isn't a "Not Found" error, return
+	// a ServerInternal problem since this is unexpected.
+	if err != nil && !berrors.Is(err, berrors.NotFound) {
+		wfe.stats.Inc("Errors.UnableToGetAccountByID", 1)
+		logEvent.AddError(fmt.Sprintf("Error calling SA.GetRegistration: %s", err.Error()))
+		return nil, nil, probs.ServerInternal(fmt.Sprintf(
+			"Error retreiving account %q", accountURL))
+	} else if berrors.Is(err, berrors.NotFound) {
+		// If the account isn't found, return a suitable problem
+		wfe.stats.Inc("Errors.KeyIDNotFound", 1)
+		logEvent.AddError(fmt.Sprintf("Account %q not found", accountURL))
+		return nil, nil, probs.AccountDoesNotExist(fmt.Sprintf(
+			"Account %q not found", accountURL))
+	}
+
+	// Verify the account is not deactivated
+	if features.Enabled(features.AllowAccountDeactivation) && account.Status != core.StatusValid {
```

Comment:
I don't think we need to keep the feature check here, v2 should always allow deactivation.